src/pyodide/internal/workers.py

```diff
@@ -682,6 +719,138 @@ async def text(self) -> str:
+            return datetime.datetime.fromtimestamp(obj.getTime() / 1000)
+        elif obj.constructor.name == "Error":
+            return Exception(obj.toString())
+        elif obj.constructor.name == "Number":
+            return obj.valueOf()
+
+        return obj
+
+    if not hasattr(obj, "constructor"):
+        return obj
+
+    result = obj.to_py(default_converter=convert)
+
+    if isinstance(result, JsProxy):
+        raise TypeError(
+            f"Couldn't convert object to Python type, got {obj.constructor.name}"
+        )
+
+    return result
+
+
+def _raise_on_disabled_type(value):
+    if isinstance(value, (tuple, bytearray)):
+        raise TypeError(f"{type(value)} cannot be sent over RPC.")
+
+    if _is_iterable(value):
+        for v in value:
+            _raise_on_disabled_type(v)
+
+
+def serialize_to_js(value) -> JsProxy:
+    """
+    Converts Python objects defined in this module (Response, Request, etc) and native Python types
+    like Map, Set, datetime to equivalent JavaScript types.
+
+    This method is used for Workers RPC in Python to convert Python objects to JavaScript. As such
+    it does not support serializing all Python object types.
+    """
+
+    def converter(obj, convert, cache):
+        if obj is None:
+            return obj
+
+        if hasattr(obj, "js_object"):
+            return obj.js_object
+
+        if isinstance(obj, datetime.datetime):
+            # TODO: Pyodide should gain support for this, we should upstream this.
+            return js.Date.new(obj.timestamp() * 1000)
+
+        if isinstance(obj, Exception):
+            return js.Error.new(str(obj))
+
+        _raise_on_disabled_type(obj)
+
+        return obj
+
+    # `to_js` won't always call the default_converter, for example when a list of tuples is passed
+    _raise_on_disabled_type(value)
+
+    result = to_js(value, default_converter=converter, dict_converter=js.Map.new)
+
+    return result
+
+
+class _FetcherWrapper:
+    def __init__(self, binding):
+        self._binding = binding
+
+    def _getattr_helper(self, name):
+        attr = getattr(self._binding, name)
+        if name == "fetch":
+            # TODO: Implement python native fetch for bindings.
+            return attr
+
+        if callable(attr):
+
+            async def wrapper(*args, **kwargs):
+                js_args = [serialize_to_js(arg) for arg in args]
+                js_kwargs = {k: serialize_to_js(v) for k, v in kwargs.items()}
+                result = await attr(*js_args, **js_kwargs)
+                return serialize_to_python(result)
+
+            return wrapper
+
+        return attr
+
+    def __getattr__(self, name):
+        result = self._getattr_helper(name)
+        setattr(self, name, result)
+        return result
+
+
+class _EnvWrapper:
+    def __init__(self, env):
+        self._env = env
+
+    def __getattr__(self, name):
+        binding = getattr(self._env, name)
+        if binding.constructor.name == "Fetcher":
```

Comment:
Can't this fail if the environment variable value is a string?

Reply:
This should have a different constructor name then, but will add a test for this.

Reply:
I think if `binding` is a string, then `binding.constructor` will raise an `AttributeError`.